Fixes part of #4374: Update docstrings in the python backend code.

YashJipkate opened the pull request:
Updated docstrings in Python Backend code for core.storage.base_model.gae_models_test, core.storage.base_model.gae_models, core.storage.user.gae_models, core.tests.test_utils, core.tests.performance_tests.base, core.tests.performance_framework.perf_services, core.tests.performance_framework.perf_domain, core.domain.summary_services, core.domain.email_jobs_one_off_test, core.domain.visualization_registry

<!--
  - Thanks for submitting code to Oppia! Please fill out the following as part of
  - your pull request so we can review your code more easily.
  -->

## Explanation
Fixes part of #4374
<!--
  - Explain what your PR does. If this PR fixes an existing bug, please include
  - "Fixes #bugnum:" in the explanation so that GitHub can auto-close the issue
  - when this PR is merged.
  -->

## Checklist
- [x] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] The PR explanation includes the words "Fixes #bugnum: ..." (or "Fixes part of #bugnum" if the PR only partially fixes an issue).
- [x] The linter/Karma presubmit checks have passed.
  - These should run automatically, but if not, you can manually trigger them locally using `python scripts/pre_commit_linter.py` and `bash scripts/run_frontend_tests.sh`.
- [x] The PR is made from a branch that's **not** called "develop".
- [x] The PR follows the [style guide](https://github.com/oppia/oppia/wiki/Coding-style-guide).
- [x] The PR is assigned to an appropriate reviewer.
  - If you're a new contributor, please ask on [Gitter](https://gitter.im/oppia/oppia-chat) for someone to assign a reviewer.
  - If you're not sure who the appropriate reviewer is, please assign to the issue's "owner" -- see the "talk-to" label on the issue.

YashJipkate committed:
Fixes part of #4374

Updated docstrings in Python Backend code for core.storage.base_model.gae_models_test, core.storage.base_model.gae_models, core.storage.user.gae_models, core.tests.test_utils, core.tests.performance_tests.base, core.tests.performance_framework.perf_services, core.tests.performance_framework.perf_domain, core.domain.summary_services, core.domain.email_jobs_one_off_test, core.domain.visualization_registry

vibhor98 commented on core/domain/email_jobs_one_off_test.py:
```diff
@@ -45,6 +45,17 @@ def test_hashes_get_generated(self):
         def _generate_hash_for_tests(
                 cls, recipient_id, email_subject, email_body):
+            """Generates hash for tests.
+            Args:
```
Please add one blank line before `Args` and `Returns` section as it was in your previous PR. Ditto everywhere else.

YashJipkate replied:
Ah OK! Didn't know that was a convention too..
Thanks!

vibhor98 commented on core/domain/summary_services.py:
```diff
@@ -493,6 +493,12 @@ def get_library_groups(language_codes):
 
     def _generate_query(categories):
+        """Generates query based on the categories and language codes.
+        Args:
```
Ditto. Leave one blank line.

YashJipkate replied:
Done!

vibhor98 replied:
Still this comment is not completely resolve. PTAL!

vibhor98 commented on core/storage/user/gae_models.py:
```diff
@@ -176,6 +176,15 @@ class ExpUserLastPlaythroughModel(base_models.BaseModel):
+        Args:
+            user_id: str. The id of the user.
+            exploration_id: str. The id of the exploration.
+        Returns:
```
Add one blank line before it

YashJipkate replied:
Done!

bansalnitish commented:
HI @YashJipkate, Thanks for the PR.
Could you change the PR title as `Fixes part of #4374: Explanation` ? Thanks

vibhor98 commented on core/storage/user/gae_models.py:
```diff
@@ -176,6 +176,15 @@ class ExpUserLastPlaythroughModel(base_models.BaseModel):
+            user_id: str. The id of the user.
+            exploration_id: str. The id of the exploration.
+        Returns:
+            key: str. The generated key using user_id and exploration_id
```
Remove `key:` so it should start with `str. The generated ...`

YashJipkate replied:
Ah Yes.. Missed this one.
Thanks!

vibhor98 commented on core/storage/user/gae_models.py:
```diff
@@ -406,6 +415,15 @@ class ExplorationUserDataModel(base_models.BaseModel):
+            user_id: str. The id of the user.
+            exploration_id: str. The id of the exploration.
+        Returns:
+            key: str. The generated key using user_id and exploration_id
```
Ditto as above.

YashJipkate replied:
Done!

vibhor98 commented on core/storage/user/gae_models.py:
```diff
@@ -608,7 +635,7 @@ def create(cls, user_id, story_id):
         Returns:
             StoryProgressModel. The newly created StoryProgressModel
-            instance.
+            `instance.
```
Remove ` before instance. Instead, it should be indented by 4 more spaces.

YashJipkate replied:
My bad :(
Done! Thanks!

vibhor98 commented on core/tests/performance_framework/perf_domain.py:
```diff
@@ -272,6 +282,7 @@ def __init__(self, page_session_metrics):
         self._validate()
 
     def _validate(self):
+        """Checks whether the page metrics are in the form of a list."""
```
Maybe simply write -> `Checks whether the page metrics are in the list format.`

YashJipkate replied:
Pardon my English... :P
Done!

YashJipkate commented:
@bansalnitish Done! PTAL!

vibhor98 commented on core/tests/performance_framework/perf_services.py:
```diff
@@ -206,12 +206,29 @@ def get_page_timings_from_cached_session(self, page_url):
+            latency: int. The latency of the server. Defaults to None.
+        Returns:
+            server: server. The browsermobproxy server.
+            proxy: proxy. The proxy with the specified configuration.
```
@YashJipkate actually `return server, proxy` returns a tuple containing one term as server instance and second one as proxy. So, could you modify the type info to `tuple(server, proxy)`.

YashJipkate replied:
Wasn't aware of that... Thanks!
Done!

vibhor98 commented on core/tests/performance_framework/perf_services.py:
```diff
@@ -271,18 +288,39 @@ def _setup_driver(self, proxy=None, use_proxy=False):
+
+        Args:
+            driver: driver. The WebDriver of the current session.
+        Returns:
```
Add blank line before it

YashJipkate replied:
@vibhor98  Isn't it already there? Pardon If I'm missing something

vibhor98 replied:
I mean before `Returns` also.

vibhor98 commented on core/tests/performance_framework/perf_services.py:
```diff
@@ -334,6 +388,10 @@ def _reload_first_exploration(self, driver):
         self._wait_until_page_load_is_finished()
 
     def _reload_demo_collections(self, driver):
+        """Reload demo collections.
```
Reloads (for consistency)

YashJipkate replied:
Done!

vibhor98 commented on core/tests/performance_tests/base.py:
```diff
@@ -45,28 +45,51 @@ def setUp(self):
+
+        Args:
+            session_count: int. Number of page load sessions used to
+            collect timing metrics. Defaults to DEFAULT_SESSION_SAMPLE_COUNT.
```
Indent by 4 more spaces.

YashJipkate replied:
Done!

vibhor98 commented on core/tests/performance_tests/base.py:
```diff
@@ -79,6 +102,12 @@ def _record_average_page_timings_from_uncached_session(
+
+        Args:
+            session_count: Number of page load sessions used to
+            collect timing metrics. Defaults to DEFAULT_SESSION_SAMPLE_COUNT.
```
Indent by 4 more spaces

YashJipkate replied:
Done!

vibhor98 commented on core/tests/performance_tests/base.py:
```diff
@@ -90,6 +119,13 @@ def _record_average_page_timings_from_cached_session(
+        Args:
+            page_config: dict. The page configuration parameters.
+            append_username: bool. Whether to append username to the
+            page URL.
```
Indent?

YashJipkate replied:
Sorry missed it.
Done!

vibhor98 commented on core/tests/performance_tests/base.py:
```diff
@@ -112,27 +148,39 @@ def _set_page_config(self, page_config, append_username=False):
                 self.page_url, self.username)
 
     def _test_total_page_size(self):
+        """Checks whether the total page size is under limit of uncached
```
under the limit of ...

YashJipkate replied:
Done!

vibhor98 commented on core/tests/performance_tests/base.py:
```diff
@@ -112,27 +148,39 @@ def _set_page_config(self, page_config, append_username=False):
             self.size_limit_uncached_bytes)
 
     def _test_total_page_size_for_cached_session(self):
+        """Checks whether the total page size is under limit of cached
```
Ditto

YashJipkate replied:
Done!

vibhor98 commented on core/tests/test_utils.py:
```diff
@@ -558,6 +558,26 @@ def delete_json(self, url, params='', expect_errors=False,
+            app: WSGI application. The WSGI application which receives the request
+                and produces response.
+            url: str. The URL to send the POST request to.
+            data: To be put in the body of the request. If params is an iterator,
```
In type info, you can add `*` which indicates that this arguments has varying types.

YashJipkate replied:
Thanks for the knowledge!!
Done :+1:

vibhor98 commented on core/tests/test_utils.py:
```diff
@@ -558,6 +558,26 @@ def delete_json(self, url, params='', expect_errors=False,
+            headers: dict. Extra headers to send.
+        
+        Returns:
+            json_response: webtest.TestResponse instance. The response of the
```
Remove var name here (`json_response`)

YashJipkate replied:
Also is it a general practice to not mention variable name in the return?
Thanks!
Done!

vibhor98 replied:
Yeah!

YashJipkate replied:
Oh Thanks for the info!

vibhor98 commented on core/tests/test_utils.py:
```diff
@@ -558,6 +558,26 @@ def delete_json(self, url, params='', expect_errors=False,
+                with webtest.forms.Upload fields included.
+            expect_errors: bool. Whether errors are expected.
+            expected_status_int: int. The expected status code.
+            upload_files: list. A list of (fieldname, filename, file_content).
```
Add type info of list contents as well. Like `list(str)`.

YashJipkate replied:
Done!

ankita240796 commented on core/tests/performance_framework/perf_services.py:
```diff
@@ -334,6 +388,10 @@ def _reload_first_exploration(self, driver):
 
     def _reload_demo_collections(self, driver):
+        """Reload demo collections.
+        Args:
```
Add a newline here.

vibhor98 commented on core/tests/test_utils.py:
```diff
@@ -558,6 +558,26 @@ def delete_json(self, url, params='', expect_errors=False,
+            upload_files: list. A list of (fieldname, filename, file_content).
+                You can also use just (fieldname, filename) and the file contents
+                will be read from disk.
+            headers: dict. Extra headers to send.
```
Ditto. Write type info like `dict(str, *)` for both key-value pairs.

YashJipkate replied:
Done!

ankita240796 commented on core/tests/performance_framework/perf_services.py:
```diff
@@ -334,6 +388,10 @@ def _reload_first_exploration(self, driver):
         self._wait_until_page_load_is_finished()
 
     def _reload_demo_collections(self, driver):
+        """Reload demo collections.
```
Reload --> Reloads

vibhor98 reviewed (changes requested):
Hi @YashJipkate, Took a pass, PTAL!

ankita240796 commented on core/tests/performance_framework/perf_services.py:
```diff
@@ -355,20 +418,45 @@ def _create_exploration(self, driver):
         return new_exploration_url
 
     def _setup_login(self, driver):
+        """Sets up user login
```
Add a fullstop here -> `Sets up user login.`

YashJipkate replied:
Thanks for the careful observation!
Done!

ankita240796 commented on core/domain/email_jobs_one_off_test.py:
```diff
@@ -45,6 +45,17 @@ def test_hashes_get_generated(self):
         def _generate_hash_for_tests(
                 cls, recipient_id, email_subject, email_body):
+            """Generates hash for tests.
+            Args:
```
Add a newline.

ankita240796 commented on core/tests/performance_tests/base.py:
```diff
@@ -79,6 +102,12 @@ def _record_average_page_timings_from_uncached_session(
+        """Record average page timings from cached session.
+
+        Args:
+            session_count: Number of page load sessions used to
```
Is this sentence complete?

ankita240796 commented on core/tests/test_utils.py:
```diff
@@ -558,6 +558,26 @@ def delete_json(self, url, params='', expect_errors=False,
             self, app, url, data, expect_errors=False, expected_status_int=200,
             upload_files=None, headers=None):
+        """Sends a post request with the data provided to the url specified.
+        Args:
```
Add a newline.

ankita240796 reviewed (changes requested):
Just a minor nit @YashJipkate, Please fix it. Thanks!

YashJipkate committed:
Review checks

YashJipkate committed:
Review checks #2

YashJipkate committed:
Review checks #3

YashJipkate committed:
Review checks #4

vibhor98 commented on core/storage/user/gae_models.py:
```diff
@@ -176,6 +176,16 @@ class ExpUserLastPlaythroughModel(base_models.BaseModel):
+        Args:
+            user_id: str. The id of the user.
+            exploration_id: str. The id of the exploration.
+        Returns:
```
This doesn't seem to be done. Add blank line before Returns

YashJipkate replied:
Done! Sorry for missing it.

vibhor98 commented on core/storage/user/gae_models.py:
```diff
@@ -406,6 +418,15 @@ class ExplorationUserDataModel(base_models.BaseModel):
     def _generate_id(cls, user_id, exploration_id):
+        """Generates key for the instance of ExplorationUserDataModel class in
+        the required format with the arguments provided.
+        Args:
```
Ditto.

YashJipkate replied:
Done!

vibhor98 commented on core/storage/user/gae_models.py:
```diff
@@ -406,6 +418,15 @@ class ExplorationUserDataModel(base_models.BaseModel):
+        Args:
+            user_id: str. The id of the user.
+            exploration_id: str. The id of the exploration.
+        Returns:
```
Ditto

YashJipkate replied:
Done!

vibhor98 commented on core/tests/performance_framework/perf_domain.py:
```diff
@@ -173,6 +173,16 @@ def get_total_page_size_bytes(self):
 
     def _get_duration_millisecs(self, event_end, event_initial):
+        """Get page load duration.
+        Args:
```
Ditto

YashJipkate replied:
Done!

vibhor98 commented on core/tests/performance_framework/perf_services.py:
```diff
@@ -206,12 +206,30 @@ def get_page_timings_from_cached_session(self, page_url):
     def _setup_proxy_server(self, downstream_kbps=None, upstream_kbps=None,
                             latency=None):
+        """Sets up a browsermobproxy server.
+        Args:
```
Ditto

YashJipkate replied:
Done!

vibhor98 commented on core/tests/performance_framework/perf_services.py:
```diff
@@ -294,6 +334,10 @@ def _is_current_user_logged_in(self, driver):
 
     def _login_user(self, driver):
+        """Logs in a user and gets the cookie of the current session.
+        Args:
```
Ditto. Add blank line before Args.

YashJipkate replied:
Done!

vibhor98 commented on core/tests/performance_tests/base.py:
```diff
@@ -79,6 +103,13 @@ def _record_average_page_timings_from_uncached_session(
 
     def _record_average_page_timings_from_cached_session(
             self, session_count=DEFAULT_SESSION_SAMPLE_COUNT):
+        """Record average page timings from cached session.
```
Records (for consistency).

YashJipkate replied:
Done!

vibhor98 reviewed (changes requested):
@YashJipkate there are few minor nits that you missed. PTAL, else LGTM! Thanks!

YashJipkate committed:
Review checks #5

Minor fixes

YashJipkate committed:
Review Fixes #5

Minor fix

YashJipkate committed:
Merge branch 'docstrings-backend' of https://github.com/YashJipkate/oppia into docstrings-backend

YashJipkate committed:
Review checks #5

Minor fixes

Merge branch 'docstrings-backend' of https://github.com/YashJipkate/oppia into docstrings-backend

vibhor98 reviewed (approved):
LGTM, thanks @YashJipkate! 👍

YashJipkate commented:
@vibhor98 Thanks for your time and guidance and the approval!

YashJipkate commented:
@ankita240796 Thanks for your review!!

vibhor98 commented:
Hi @apb7, Can you review this PR also being one of the code owners? Thanks!

apb7 commented on core/domain/email_jobs_one_off_test.py:
```diff
@@ -45,6 +45,19 @@ def test_hashes_get_generated(self):
+            Returns:
+                str. Empty if recipient_id is 'recipient_id2',
+                    None if 'recipient_id1' and 'Email Hash'
+                    otherwise.
```
Can we shift `otherwise` to the above line without exceeding the 80 characters limit? If yes, then please go ahead and shift it.

YashJipkate replied:
Done!

apb7 commented on core/storage/base_model/gae_models.py:
```diff
@@ -429,6 +430,7 @@ def _compute_snapshot(self):
         return self.to_dict(exclude=['created_on', 'last_updated'])
 
     def _reconstitute(self, snapshot_dict):
+        """Populates the model instance with the snapshot."""
```
Hi @vibhor98, don't we require `Args` for `snapshot_dict` here?

YashJipkate replied:
Hi @apb7 I have added it for now; will remove if turns out to be unnecesary.
Also, I am skeptical about the way I wrote the return type. PTAL. Thanks!

vibhor98 replied:
It's optional in test files. Else, we need to write them. @YashJipkate, thanks for adding them here!

apb7 commented on core/tests/performance_tests/base.py:
```diff
@@ -45,28 +45,52 @@ def setUp(self):
 
     def _record_page_metrics_from_uncached_session(self):
+        """Records the page metrics from uncached session for a given page
+        URL.
```
Can we move `URL` to the above line as well?

YashJipkate replied:
Actually it exceeds the limit in this case...
Thanks!

apb7 commented on core/tests/performance_tests/base.py:
```diff
@@ -45,28 +45,52 @@ def setUp(self):
 
     def _record_page_metrics_from_cached_session(self):
+        """Records the page metrics from cached session for a given page
+        URL.
```
Ditto.

YashJipkate replied:
@apb7 Done!!

apb7 commented on core/tests/performance_tests/base.py:
```diff
@@ -90,6 +121,13 @@ def _record_average_page_timings_from_cached_session(
+        Args:
+            page_config: dict. The page configuration parameters.
+            append_username: bool. Whether to append username to the
+                page URL.
```
Ditto.

YashJipkate replied:
@apb7 Done!!

apb7 reviewed (commented):
Hi @YashJipkate, this looks good! Left a few comments. Also, can you please ensure that we do not cutoff any line before the 80 character limit? Thanks!

YashJipkate committed:
Moved a line up where it can

YashJipkate committed:
Merge branch 'docstrings-backend' of https://github.com/YashJipkate/oppia into docstrings-backend

YashJipkate committed:
linting checks

linting checks ~2

linting checks ~3

YashJipkate committed:
Added missing args and return docstrings

vibhor98 commented on core/storage/base_model/gae_models.py:
```diff
@@ -429,6 +430,16 @@ def _compute_snapshot(self):
+                property values.
+
+        Returns:
+            self: the class instance. The instance of the model class
```
@YashJipkate, here type info is wrong. Instead it should be `VersionedModel ` since it's the instance of the class. Also, remove `self` as we don't write var in returns.

YashJipkate replied:
@vibhor98 Made some changes. PTAL!
Actually, I thought that the `self` was not really a variable, since it refers to the model instance and is a keyword and also wasn't initialized anywhere. That's why I mentioned it in the return statement.
WIll keep that in mind the next time. Thanks!

YashJipkate commented:
@apb7 PTAL. Thanks!

YashJipkate committed:
Address Review

YashJipkate committed:
Linting fix

Linting fixes

vibhor98 reviewed (approved):
LGTM, thanks @YashJipkate! @apb7, can you take a final look? Thanks!

YashJipkate commented:
@apb7 just a gentle reminder. Kindly review.
Thanks!

apb7 reviewed (approved):
LGTM, thanks @YashJipkate!

vibhor98 commented:
Hi @BenHenning, can you review this PR being one of the code owners? Thanks!

YashJipkate commented:
Hi @BenHenning, just a gentle reminder. Kindly review.
Thanks!

YashJipkate commented:
Hi @BenHenning, A reminder again. Kindly review.
Thanks!

vibhor98 commented:
Hi @seanlip, can you review this PR once since you're one of the code owners? Thanks!

seanlip commented on core/tests/performance_framework/perf_domain.py:
```diff
@@ -173,6 +173,18 @@ def get_total_page_size_bytes(self):
+        """Get page load duration.
+
+        Args:
+            event_end: str. The event upto which duration measurement is
```
Nit: upto --> up to

YashJipkate replied:
Done!

seanlip commented on core/tests/performance_framework/perf_services.py:
```diff
@@ -206,12 +206,31 @@ def get_page_timings_from_cached_session(self, page_url):
+            downstream_kbps: int. The downstream speed in kbps. Defaults to
+                None.
+            upstream_kbps: int. The upstream speed in kbps. Defaults to None.
+            latency: int. The latency of the server. Defaults to None.
```
In what unit?

YashJipkate replied:
It is in milliseconds.
Done!

seanlip commented on core/tests/performance_framework/perf_services.py:
```diff
@@ -206,12 +206,31 @@ def get_page_timings_from_cached_session(self, page_url):
+            latency: int. The latency of the server. Defaults to None.
+
+        Returns:
+            tuple(server, proxy): A tuple consisting of the server and the
```
I don't think "server" and "proxy" are the actual type names; ditto for "driver" below and throughout the remainder of this file. Could you please make sure the types are correct? For instances of classes it should be the class name (so, in this case, it should be Server for the first arg, since it's an instance of browsermobproxy.Server).

Don't just change everything to start with uppercase, though -- make sure to actually figure out what class it is an instance of. When resolving this comment, please also include (in your comment) a reference to where the driver class is defined.

Thanks!

YashJipkate replied:
The driver class is defined as [here](https://github.com/SeleniumHQ/selenium/blob/master/py/selenium/webdriver/firefox/webdriver.py#L40)
Changed Server type name and updated proxy type to dict.
PTAL. Thanks!

seanlip commented on core/tests/test_utils.py:
```diff
@@ -558,6 +558,27 @@ def delete_json(self, url, params='', expect_errors=False,
+            headers: dict(str, *). Extra headers to send.
+        
+        Returns:
+            webtest.TestResponse instance: The response of the
```
webtest.TestResponse. The response of the POST request.

YashJipkate replied:
Done.

seanlip reviewed (commented):
Hi @YashJipkate, thanks for this PR. Just a few comments but otherwise it looks great!

seanlip reviewed (approved):
LGTM. Thanks!

YashJipkate commented on core/tests/test_utils.py:
```diff
@@ -555,9 +555,30 @@ def delete_json(self, url, params='', expect_errors=False,
             upload_files=None, headers=None):
+        """Sends a post request with the data provided to the url specified.
+
+        Args:
```
@seanlip I had to add the disable pylint statement [here](https://github.com/oppia/oppia/pull/6028/files/03db4263c2eb34f8771750fcff58dbfe04cb99c9#diff-6f3c73d302248e02f5b3eeb202eea43dR558) because it was saying `missing-param-doc` for parameter "app" but it was clearly there. Also I noticed that changing `Args` to `Params` solved the issue but I didnt went ahead with it as it would break the style. PTAL!
Thanks.

YashJipkate commented:
@nithusha21 @seanlip I just saw it is showing 140 changed files changed. I recently pulled the latest develop branch in my current `docstrings-backend` branch which has caused this. But the changes in all the files except the 10 I have changed myself are already there in the main develop branch of oppia. I even ran `git diff` between this branch and the main develop branch of Oppia and found the changes in only the 10 files I modified. I have experienced this same issue earlier and think this is some kind of bug in github's method of displaying diff in the PR where it doesn't compare correctly with the current master branch. PTAL and let me know if there are any suggestions regarding this.

seanlip commented:
Hi @YashJipkate, I think @apb7 is the person you'll need to talk to for both these issues. The pylint-disable statement should not be added; instead, please try to figure out _why_ missing-param-doc is arising erroneously. If it works correctly for other functions but not for yours, then it might actually be a problem with your code. (In general, don't use pylint pragmas and other similar "escape hatches" to get out of understanding an issue.)

For the 140 files changed, I think this is a dev workflow issue and that @apb7 or @nithusha21 might be able to advise. But, generally, if you follow the steps on the [PR instructions page](https://github.com/oppia/oppia/wiki/Contributing-code-to-Oppia#instructions-for-making-a-code-change), I believe that this shouldn't happen.

Note that @apb7 is currently travelling so you might want to try and take a stab at resolving these on your own first. For the second one it's fine to create a new PR from scratch if that is easier.

YashJipkate committed:
Reviews and linting

YashJipkate committed:
Fixed missing params error

YashJipkate commented on core/tests/test_utils.py:
```diff
@@ -560,8 +560,8 @@ def _send_post_request(
-        Params:
-            app: WSGI application. The WSGI application which receives the
+        Args:
+            app: TestApp. The WSGI application which receives the
```
@seanlip Thanks for your advice! Will keep that in mind next time!
Well thanks to you, I figured out the cause, it was caused since I named the type as `WSGI Application` and pylint was not happy with the space in between the type. So I set the type to `TestApp` which is the class as is clear from the piece of code [here](https://github.com/oppia/oppia/blob/develop/core/tests/test_utils.py#L594) and also from the [source code of TestApp](https://github.com/Pylons/webtest/blob/master/webtest/app.py#L89)
PTAL. Thanks!

seanlip commented:
@YashJipkate, the PR looks good, but I noticed you force pushed. We generally discourage that because it overwrites history and makes diffs between successive commits harder to review. Instead perhaps reverting the merge commit (or whatever caused all the new files to be added) might have been a better way to go -- though it might be a good idea to try and figure out how you got into this state in the first place, so that we can avoid this happening to other folks in the future.

Otherwise, the PR looks good, except that it fails Travis lint checks. Did these not run locally on your machine before you pushed?

Once the Travis checks are fixed, this is good to merge. Thanks!

YashJipkate committed:
Linting checks

YashJipkate commented:
@seanlip Thanks for the info. I actually used `git reset` and then force-pushed my repo (as you saw) at a previous time at a different repo when I encountered this issue, and they were fine with it so I went ahead with it here too.
Wasn't aware of the hardship it causes to reviewers.
And the linting check did run on my local since I used the terminal, don't know how it passed there, might be because my local branch is behind `oppia/develop`. I have pushed again with the fix.
Thanks!

seanlip commented:
OK, thanks. No worries. 

I'm a bit concerned, though, that this issue seems to be recurring repeatedly for you. If it happens in the future can you document exactly what steps you took locally and file a bug against the dev workflow team so that they can look into it?

YashJipkate commented:
@seanlip Yeah sure I will keep that in mind the next time I encounter this type of issue.
Thanks!

apb7 commented:
@YashJipkate, glad that you've figured out the above issues! Also, as Sean suggested, feel free to file an issue if you face any recurring problem. Thanks!

YashJipkate commented:
@apb7 Sure. Thanks for your help!

YashJipkate commented:
@seanlip The checks have passed. Please take a final look.
Thanks!

seanlip commented:
LGTM. Thanks!
